### PR TITLE
[SPARK-44507][SQL][CONNECT] Move AnalysisException to sql/api

### PR DESCRIPTION
**Pull request by hvanhovell**
### What changes were proposed in this pull request?
This PR moves `AnalysisException` to sql/api. In order to do so we had to move the `plan` field out. We have introduced the `ExtendedAnalysisException` subclass to host this field.

### Does this PR introduce _any_ user-facing change?
Yes, `AnalysisException` does not have the `plan` field anymore. This is a breaking change. On the upside this reduces a lot of the breaking changes we had to make, for example `ParseException` is an `AnalysisException` again.

### How was this patch tested?
(Updated) Existing tests.

### CODE REVIEW

**Comment by hvanhovell** [comment]
If we go with this approach we should undo some of the error work in sql/api that uses SparkException now instead of AnalysisException.

**Comment by hvanhovell** [comment]
cc @amaliujia @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ExtendedAnalysisException.scala [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+package org.apache.spark.sql.catalyst
+
+import org.apache.spark.QueryContext
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+
+/**
+ * Internal [[AnalysisException]] that also captures a [[LogicalPlan]].
+ */
+class ExtendedAnalysisException(
```
Comment (zhenlineo) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ExtendedAnalysisException.scala:26]: This means users would see the new error class in stacktraces? e.g.
`org.apache.spark.sql.catalyst.AnalysisException` -> `org.apache.spark.sql.catalyst.ExtendedAnalysisException`

Should we try to keep the original name by overriding `toString` method?

```
Throwable.java
public String toString() {
        String s = getClass().getName();
        String message = getLocalizedMessage();
        return (message != null) ? (s + ": " + message) : s;
    }
```

Comment (hvanhovell) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ExtendedAnalysisException.scala:26]: I would be OK with that.

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ExtendedAnalysisException.scala (hunk 1 above)
Comment (amaliujia) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ExtendedAnalysisException.scala:26]: The current idea is there are a bit less place to use the plan field of the `AnalysisException` thus this switch cause less changes.

**Review by amaliujia — APPROVED** [review]

**Review by heyihong — APPROVED** [review]
:+1:

**Comment by hvanhovell** [comment]
Merging this. Tests failures are known to be problematic/flaky.

**Comment by dongjoon-hyun** [comment]
Hi, @hvanhovell.

Is this JIRA ID correct? SPARK-44507 seems to be resolved as `Won't Fix` with no `Fix Version`. Could you update properly?

https://issues.apache.org/jira/browse/SPARK-44507

Screenshot 2024-05-09 at 13 35 07

**Pull request closed without merging**